llama : fix segfault from unknown model arch name

compilade opened the pull request:
As described in <https://github.com/ggerganov/llama.cpp/pull/5795#discussion_r1509222880>, when comparing the model arch name, if `llm_arch_from_string` gets to `LLM_ARCH_UNKNOWN` in the `LLM_ARCH_NAMES` map, a NULL pointer for the name is obtained. This case needs to be handled to avoid a segmentation fault.

To test that this works, load a model with a name that is not in `LLM_ARCH_NAMES` (an easy way is to do this is to change the names in `LLM_ARCH_NAMES` (e.g. changing "llama" to "lllama")).

With `master`, it segfaults. With this PR, it correctly errors with
```
llama_model_load: error loading model: error loading model architecture: unknown model architecture: 'whatever-unknown-name-you-tried'
```

compilade committed:
llama : fix segfault from unknown model arch name

slaren commented:
I think we need to stop using the `operator[]` here to avoid adding unknown architectures to the list:
https://github.com/ggerganov/llama.cpp/blob/c2224f003bf9cf558b1a3c57033563e11a4de9a5/llama.cpp#L364-L366

cebtenzzre commented:
> I think we need to stop using the `operator[]` here to avoid adding unknown architectures to the list:

Yep, this should use std::map::at instead.

compilade commented:
Hmm, using `std::map::at` there instead errors for every model type (even known ones) with
```
llama_model_load: error loading model: map::at
llama_load_model_from_file: failed to load model
```
This happens because the `operator()` of `llm_kv` is first called with `LLM_ARCH_UNKNOWN` to get the architecture name, but it's not in `LLM_ARCH_NAMES`.

Maybe the way to fix this would be to define `LLM_ARCH_UNKNOWN` to have an empty string as a name?

ggerganov commented:
And also make all LLM maps `const`

cebtenzzre commented:
> Maybe the way to fix this would be to define `LLM_ARCH_UNKNOWN` to have an empty string as a name?

I think the only acceptable name for LLM_ARCH_UNKNOWN is something like `unknown` or `(unknown)` - in case it ever gets printed.

compilade committed:
llama : make all LLM maps const

This also requires using `std::map::at` instead of its `operator[]`
which does not exist for const maps.

* llama : name LLM_ARCH_UNKNOWN to "(unknown)"

This avoids errors from `std::map::at` when
getting the general name of the model architecture.
Using "(unknown)" instead of an empty string as per suggestion
https://github.com/ggerganov/llama.cpp/pull/5820#issuecomment-1973735284

cebtenzzre commented on llama.cpp:
```diff
@@ -397,7 +398,7 @@ enum llm_tensor {
 };
 
-static std::map<llm_arch, std::map<llm_tensor, std::string>> LLM_TENSOR_NAMES = {
+static const std::map<llm_arch, const std::map<llm_tensor, std::string>> LLM_TENSOR_NAMES = {
```
The extra const won't do anything here as const maps return const references to values.
```suggestion
static const std::map<llm_arch, std::map<llm_tensor, std::string>> LLM_TENSOR_NAMES = {
```

compilade committed:
llama : remove redundant inner const for LLM_TENSOR_NAMES

The extra const won't do anything here as const maps
return const references to values.

Co-authored-by: Jared Van Bortel <cebtenzzre@gmail.com>

compilade committed:
llama : remove redundant nullptr check in llm_arch_from_string

Since LLM_ARCH_NAMES is a const map, no spurious elements
with a NULL name are inserted anymore, so this check is dead code.
